src/_pytest/pathlib.py

```diff
@@ -689,6 +735,60 @@ def resolve_package_path(path: Path) -> Optional[Path]:
     return result
 
 
+def resolve_pkg_root_and_module_name(
+    path: Path, *, consider_ns_packages: bool = False
+) -> Tuple[Path, str]:
+    """
+    Return the path to the directory of the root package that contains the
+    given Python file, and its module name:
+
+        src/
+            app/
+                __init__.py
+                core/
+                    __init__.py
+                    models.py
+
+    Passing the full path to `models.py` will yield Path("src") and "app.core.models".
+
+    If consider_ns_packages is True, then we additionally check upwards in the hierarchy
+    until we find a directory that is reachable from sys.path, which marks it as a namespace package:
+
+    https://packaging.python.org/en/latest/guides/packaging-namespace-packages
+
+    Raises CouldNotResolvePathError if the given path does not belong to a package (missing any __init__.py files).
+    """
+    pkg_path = resolve_package_path(path)
+    if pkg_path is not None:
+        # https://packaging.python.org/en/latest/guides/packaging-namespace-packages/
+        pkg_root = pkg_path.parent
+        if consider_ns_packages:
+            # Go upwards in the hierarchy, if we find a parent path included
+            # in sys.path, it means the package found by resolve_package_path()
+            # actually belongs to a namespace package.
+            for parent in pkg_path.parents:
```

Comment:
This loop is likely to be quite slow, and it also run for prepend/append modes. Any way to make it faster?

Reply:
Nothing occurs to me at the moment.

Perhaps we should make `consider_namespace_packages` an ini-option? I think this is a good idea for two reasons:

1. I fear this might cause some pretty bad regressions in the wild, having this feature behind an option would prevent regressions from users not using/interested in namespace packages.
2. It incurs a performance penalty, for the reasons we discussed: it is easy for Python to go from module_name -> path, the other way around requires more work.


What do you think?

Reply:
Will it default to true or false?

Reply:
To `false`, I think; seems better for users using namespace package to set the option (and pay the performance penalty for it) rather than enabling this for everyone.

Reply:
Sounds good to me, maybe we can flip it to true in the future.

Reply:
BTW, another part of namespace pkg support is to reflect it in the collection tree as `Package` (or `NamespacePackage`?) nodes. But that's a different matter.

Reply:
I will add the option/docs then.

> Also for your consideration, I've split the PR into commits, I think it will help with reviewing/bisecting/reverting compared to one big commit. https://github.com/bluetech/pytest/commits/11475-importlib/

That looks great, can you push force those commits into my branch? Then I can continue from there.

Reply:
Done!